文件验证 -by- 徐梦旗 201732120124

remeio opened the pull request:
修改isGene1OrGene2Json函数；
描述：
上传文件时，对文件内容是否为JSON进行验证。 文件是合格JSON文件，但不一定是合法的gene1或gene2内容， 比如合法JSON文件的内容： { "R0SRR1738120XX":{"value":1.71018781553424}, "R0SRR1738121XX":{"value":1.81645208181843} } 但不是合法的gene1内容，合法的内容应该是： { "R0SRR1738120XX":1.71018781553424, "R0SRR1738121XX":1.81645208181843 }

remeio committed:
文件验证 -by- 徐梦旗 201732120124

修改isGene1OrGene2Json函数；
描述：
上传文件时，对文件内容是否为JSON进行验证。 文件是合格JSON文件，但不一定是合法的gene1或gene2内容， 比如合法JSON文件的内容： { "R0SRR1738120XX":{"value":1.71018781553424}, "R0SRR1738121XX":{"value":1.81645208181843} } 但不是合法的gene1内容，合法的内容应该是： { "R0SRR1738120XX":1.71018781553424, "R0SRR1738121XX":1.81645208181843 }

spm2020spring commented:
有道理。

@xumengqi1 测试呢？

@bananahab 能不能Review一下？

@BaekLi 有什么看法吗？

蓝珲

remeio commented:
测试由王雪洁负责，她正在编写测试用例，并将于今天提交测试用例。



---原始邮件---
发件人: "Software Project Management"<notifications@github.com&gt;
发送时间: 2020年4月2日(周四) 晚上8:01
收件人: "lanlab-org/GeneExpressionScatterPlot-XuMengqi"<GeneExpressionScatterPlot-XuMengqi@noreply.github.com&gt;;
抄送: "xumengqi1"<2663479778@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题: Re: [lanlab-org/GeneExpressionScatterPlot-XuMengqi] 文件验证 -by- 徐梦旗 201732120124 (#6)




 
有道理。
 
@xumengqi1 测试呢？
 
@bananahab 能不能Review一下？
 
蓝珲
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

spm2020spring commented on gene_expression_scatter_plot/js/main.js:
```diff
@@ -766,18 +766,39 @@ function makeNumberToStringAndExponential(temp) {
+ * @param {Object} str 合法JSON格式的字符串
+ * @return {bool} 是否为合法的Gene1和Gene2内容
+ */
 function isGene1OrGene2Json(str) {
```
function **isValidGeneExpressionData**(str) {

spm2020spring replied:
function **isValidGeneExpressionData**(str) {

spm2020spring commented on gene_expression_scatter_plot/js/main.js:
```diff
@@ -766,18 +766,39 @@ function makeNumberToStringAndExponential(temp) {
-        return true;
-    }
-    return false;
+	//valid json
+	//{string : number, string : number} ok
+	//{string : {string : number}} not ok
+	//{string : number, string : [number1, number2]} not ok
+	//[number, number, number] not ok
+	//number not ok
+	//合法条件为：以 { 开头，以 } 结束；只有一个 {，只有一个 }；不存在 [ 或 ]；
+	var string = str.trim();
+	var isStartOk = (string.charAt(0) == '{');
+	var isEndOk = (string.charAt(string.length - 1) == '}');
+	var isJustOneLeft = true;
+	var isJustOneRight = true;
+	var isJustLeftAndRight = true;
+    for (var i = 1; i < str.length - 1; i++) {
+		var aChar = string.charAt(i);
+        if (aChar == '{') {
+            isJustOneLeft = false;
+			break;
+        }
+		if (aChar == '}') {
+			isJustOneRight = false;
+			break;
+		}
+		if (aChar == '[' || aChar == ']') {
+			isJustLeftAndRight = false;
+			break;
+		}
+    }
+	//判断顺序不能改变，用了短路
+    return isStartOk && isEndOk && isJustOneLeft && isJustOneRight && isJustLeftAndRight;
```
@xumengqi1 

@bananahab 


没有简单一点的判断方法吗？

比如： 先判断是不是dictionary。如果不是，返回false。如是，则对于dictionary每个key，判断它对应的value是不是数值，只要不是数值，就马上返回false。

蓝珲

remeio replied:
该函数传入的参数为字符串类型，所以是按JSON文件转换成字符串后的特征进行处理；

remeio committed:
更新无效基因内容的错误码内容 -by- 徐梦旗 201732120124

报错108的本质原因是：gene1或gene2的内容无效，导致gene1或gene2的内容无效的一种情况是gene文件和information上传顺序颠倒了，所以用maybe。

remeio committed:
语句问题（不重要） -by- 徐梦旗 201732120124

an error 不是 a error。

remeio committed:
更新无效基因内容的错误码内容 -by- 徐梦旗 201732120124

remeio committed:
添加isGene1OrGene2JsonTest测试用例 -by- 王雪洁 201732120106

添加测试用例，并测试
测试效果：http://39.97.255.112/gesp/test/
